Rewrite Transaction as a class

fterh opened the pull request:
* Implement generic observer pattern
* Add test case for Transaction+Equatable
* Conform Transaction to Obserable
* Add test cases for Transaction+Observable

fterh committed:
Implement generic observer pattern protocols

fterh committed:
Update observer-observable implementation

fterh committed:
Rewrite Transaction as class

fterh committed:
Conform Transaction to Observable

fterh committed:
Add negative test case for equatable

fterh committed:
Test for Transaction + Observable logic

GilgameshTC commented on bacon/model/Transaction.swift:
```diff
@@ -42,11 +58,105 @@ struct Transaction: Codable, Equatable {
+
+extension Transaction: Equatable {
+
+    static func == (lhs: Transaction, rhs: Transaction) -> Bool {
```
wouldn't this defeat the identification purpose whereby 2 transactions of same value should be different transactions?

fterh replied:
Not really, there is the triple-equal `===` to test for identity. I guess this is useful for checking if 2 separate transactions have the same "contents", i.e. 2 different transactions that are equivalent.

Granted, I can't think of a situation where this is useful, but I thought I should leave the Equatable conformance in in case it is utilized in another part of the code.

kumuwu replied:
If the user's new value is equal to old value when he tries to edit, we can use == to check if there is even a need to notify observers. But even if we notify anyway it should be okay as well, but this can be a possible use case

fterh replied:
Resolved using didSet()

GilgameshTC commented on bacon/commons/Observer.swift:
```diff
@@ -0,0 +1,21 @@
+    /// Receives notification of a value published by the observed subject.
+    /// Since `value` is typed as `Any`, the observer is responsible for casting `value` to
+    /// the correct type of whatever it is observing.
+    func notify(_ value: Any)
```
should this be notified instead? notify sounds abit confusing haha

fterh replied:
I think notify is fine, all the observer patterns implementations I found use the imperative form of the word too haha

fterh committed:
Merge branch 'master' into transaction-class

fterh committed:
Merge branch 'master' into transaction-class

kumuwu commented on bacon/model/Transaction.swift:
```diff
@@ -42,11 +58,105 @@ struct Transaction: Codable, Equatable {
+        notifyObservers(self)
+    }
+
+    // Public computed properties
```
Is using computed properties only because you need to call set()?
If that's the case, I think using a normal property + didSet() is fine. (since didSet's primary purpose is to act as an observer anyway)
This is also because I read/heard somewhere before that if you provide getter and setter for every property you might as well make it public (but I am not 100% sure of this; I searched on this and it seems like it's debatable)

fterh replied:
Okay, done :)

kumuwu commented on baconTests/TransactionTests.swift:
```diff
@@ -61,7 +61,32 @@ class TransactionTests: XCTestCase {
+        XCTAssertEqual(observer.notifiedCount, 1)
+        transaction.amount = 2 // Set amount to same value
+        XCTAssertEqual(observer.notifiedCount, 2)
     }
```
btw, me and travis was talking about this yesterday: we can just disable force try at the start of the file and enable at the end of the file HAHA

kumuwu commented on bacon/model/Transaction.swift:
```diff
@@ -8,13 +8,29 @@
+    private var _category: TransactionCategory
+    private var _amount: Decimal
+    private var _description: String
+
```
why description? I thought we don't have description but just a picture which is NSData?

fterh replied:
I know we want an `image` property but are we replacing description? :O

fterh committed:
Rewrite computed properties as stored properties with didSet observer

fterh committed:
Disable swiftlint force_try rule for test file

Rather than having a bunch of those comments scattered throughout the code

fterh committed:
Implement CodableCLLocation wrapper

CLLocation is not Codable by default. CodableCLLocation is a Codable wrapper around CLLocation for use as a Transaction property for encoding/decoding.

fterh committed:
Fix swiftlint violations

fterh committed:
Merge branch 'codable-CLLocation' into transaction-class

fterh committed:
Add != static method for Equatable conformance
